src/vec-zsts.md

```diff
@@ -35,55 +35,57 @@ method of RawVec.
 ```rust,ignore
 impl<T> RawVec<T> {
     fn new() -> Self {
-        // !0 is usize::MAX. This branch should be stripped at compile time.
-        let cap = if mem::size_of::<T>() == 0 { !0 } else { 0 };
-
+        // This branch should be stripped at compile time.
+        let cap = if mem::size_of::<T>() == 0 {
+            std::usize::MAX
+        } else {
+            0
+        };
         // Unique::empty() doubles as "unallocated" and "zero-sized allocation"
-        RawVec { ptr: Unique::empty(), cap: cap }
+        Self {
+            ptr: Unique::empty(),
+            cap,
+        }
     }
 
     fn grow(&mut self) {
         unsafe {
-            let elem_size = mem::size_of::<T>();
-
-            // since we set the capacity to usize::MAX when elem_size is
+            // Since we set the capacity to `usize::MAX` for ZST is
             // 0, getting to here necessarily means the Vec is overfull.
-            assert!(elem_size != 0, "capacity overflow");
+            assert!(mem::size_of::<T>() != 0, "capacity overflow");
```

Comment:
You should make this check the layout.size(), but I guess that doesn't matter because you probably need to redo this code because this is the wrong layout in the general case.